Météo-france sensor : current weather and 1 hour rain forecast

victorcerutti opened the pull request:
## Description:
Météo France provides a convenient rain forecast that is pretty accurate.
This sensor gives the time to next rain forecasted.
It also display several sensor for monitored conditions, as in the example
@MartinHjelmare @fabaff : I screwed up my last PR during a rebase. I took account of your requests and this is a new release of the Météo-France platform, using a pypi package


**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>
home-assistant/home-assistant.io#6938


## Example entry for `configuration.yaml` 
```yaml
sensor:
  - platform: meteo_france
    postal_code: 76000
    monitored_conditions:
      - rain_chance
      - freeze_chance
      - thunder_chance
      - snow_chance
      - weather
      - next_rain
      - wind_speed
      - temperature
      - uv
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

victorcerutti committed:
🌧 Meteo France rain forecast

victorcerutti committed:
do not fail on wrong config

victorcerutti committed:
Update name of sensor class

victorcerutti committed:
do not show sensor if not working

victorcerutti committed:
Update .coveragerc

victorcerutti committed:
ability to submit insee location code without final 0 needed by meteo-france

victorcerutti committed:
Lynting

victorcerutti committed:
more lynting

victorcerutti committed:
update comment

victorcerutti committed:
block comment linting

victorcerutti committed:
reducing length of long lines

victorcerutti committed:
linting

victorcerutti committed:
Update météo-france platform

Now work with pypi package and monitored conditions

victorcerutti committed:
remove error log

fabaff commented:
Successor of #17602.

victorcerutti committed:
Update requirements_all.txt

victorcerutti committed:
Increase scan interval to 5 minutes

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+Support for Meteo France raining forecast.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.meteofrance/
```
There's a missing underscore in the page name.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+
+import logging
+import datetime
+import voluptuous as vol
```
Add a blank line between standard library and 3rd party imports.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_POSTAL_CODE): cv.string,
+    vol.Required(CONF_MONITORED_CONDITIONS, default=[]):
```
Default isn't necessary if key is required.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Rename `add_devices` to `add_entities`.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Meteo-France sensor."""
+    postal_code = config.get(CONF_POSTAL_CODE)
```
Use `dict[key]` for required config keys.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._data["name"]+' '+SENSOR_TYPES[self._condition][0]
```
Use new style string formatting instead of string concatenation.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+        """Return the state of the sensor."""
+        if self._data[self._condition] is not False:
+            return self._data[self._condition]
+        return STATE_UNKNOWN
```
Use `None` to represent unknown state.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+        return STATE_UNKNOWN
+
+    @property
+    def state_attributes(self):
```
Replace this with `device_state_attributes`.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,132 @@
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes of the sun."""
```
Stale docstring.

victorcerutti committed:
Update meteo_france according to review

victorcerutti committed:
better error handling of location missing some monitored conditions

victorcerutti committed:
fix lint error

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,140 @@
+        try:
+            return self._data[self._condition]
+        except KeyError:
+            _LOGGER.error("No result for the condition `%s`", self._condition)
```
Please don't log in entity properties. We don't want any side effects here.

victorcerutti replied:
Thanks for the review.
Some cities doesn't have all the conditions returned by the package. What should I do about the sensor then ?
Without the exception catching, it fails, generates an error and the sensor is not displayed on the front-end.
I can also keep the exception catching without logging. Sensors stay on the front-end with a `None` value and it's up to the user to remove the monitored-condition that is not available for this location. But without error logging, there is not hint that it's the problem

Maybe another solution would be to still catch those KeyError and set the state as "no data available for this location".

MartinHjelmare replied:
Just move the access an exception handling etc to `update` and save the state in an instance attribute. Return the instance attributes here in the `state` property.

victorcerutti replied:
latest commit should fix this

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,140 @@
+                    }
+                }
+            except KeyError:
+                _LOGGER.error("No result for the condition `%s`",
```
See above.

MartinHjelmare replied:
This isn't resolved.

victorcerutti committed:
moving error log

Errors are now catched at sensor initialization and state is set when updating the data

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,142 @@
+        self._data = client.get_data()
+        self._state = None
+        try:
+            self._state = self._data[self._condition]
```
Instead call `add_entities` with second argument `True`. It will make sure that `update` is called during entity addition.

We don't want side effects in init either.

MartinHjelmare commented on homeassistant/components/sensor/meteo_france.py:
```diff
@@ -0,0 +1,142 @@
+        """Initialize the sensor."""
+        self._condition = condition
+        self._client = client
+        self._data = client.get_data()
```
Why are we only calling this here, one time? Shouldn't this be called on every update?

We should probably move this to `update`.

victorcerutti replied:
I'm new to Python but here is my understanding : 
As `client.get_data` returns a `dict`, it passes the reference to the client data
When it's updated after the remote fetch/parsing of the package, the sensor `_data` is up to date.

I updated to call `get_data()` after the update to be sure the reference is never broken for whatever reason may happen

victorcerutti committed:
Update updating of sensor

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
@fabaff ok with you?

fabaff reviewed (approved):
Fine with me :bird:

romquenin commented:
Hello, I would love to use this sensor for it's french human-readable text summary of the current conditions but unfortunately my foreign city is not available. Could it be possible to add a european city with it postal code (previsions-meteo-monde/londres/03772)  03772 for london.
It would be a great feature for French People living abroad.
Thanks.

MartinHjelmare commented:
If you want to suggest an enhancement please open a feature request in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum.

Merged PRs should not be used for enhancement discussion. 

Thanks!
